### ISSUE TITLE
On asyncio, `Event.set()` sometimes fails to notify all waiting tasks

### ISSUE THREAD

**Issue by gschaffner** [agronholm/anyio#536]
hi!

if:

1.  task A is awaiting `Event.wait()`
1.  `Event.set()` is called
1.  the `wait()`'s cancel scope is cancelled before the event loop schedules A

then, on the asyncio backend only, task A's `wait()` raises cancelled instead of getting notified of the event.

in contrast, on the trio backend (and on pure trio), `Event.set()` still notifies all waiters (including task A) in this situation. trio does this by implementing `Event.wait` via `wait_task_rescheduled`.

here is a small reproducer: `test_event_wait_before_set_before_cancel` in 65d74e483cbcb161223bc9216b020156104f09c0. `assert wait_woke` fails on asyncio only.

this problem is also the direct cause of another bug: `MemoryObjectSendStream.send(item)` can raise cancelled after `item` has been delivered to a receiver![^1] this `MemoryObjectSendStream.send` issue is just the `send_event.wait()` case to #146's `receive_event.wait()` case, i believe—the underlying issue for both seems to be this `Event.wait()` problem[^2].

a couple relevant comments from njsmith: https://github.com/agronholm/anyio/issues/146#issuecomment-673754182, https://github.com/agronholm/anyio/pull/147#issuecomment-674615765.

[^1]: raising cancelled after a receiver gets the item is a violation of trio's cancellation semantics for API in the `trio` namespace (https://trio.readthedocs.io/en/stable/reference-core.html#cancellation-and-primitive-operations). since `anyio` is documented as following `trio`'s cancellation model, i would guess that `anyio` is also intended to adhere to this (but in the `anyio` namespace)? if so, i think it would be good to document this at https://anyio.readthedocs.io/en/stable/cancellation.html and to document the couple rare exceptions to this rule (e.g. https://trio.readthedocs.io/en/stable/reference-io.html#trio.abc.SendStream.send_all) too.

[^2]: nit: #146 actually had two underlying issues: the problem that `MemoryObjectReceiveStream.receive()` didn't `checkpoint_if_cancelled()` before trying `receive_nowait()` (on all backends), as well as this `Event.wait()` issue (on asyncio only).

**Comment by agronholm** [comment]
I'm really confused now. If a task is cancelled while blocked on `Event.wait()`, why should it not raise `CancelledError` if it's cancelled before the event loop gets around to running it? And why would it affect other tasks waiting on the same event? The subject line itself is misleading because `Event.wait()` doesn't notify any tasks, it waits for a notification. What am I missing here?

**Comment by smurfix** [comment]
My rationale for why this is a problem: Assume that the event is associated with reliably transferring some data from task A to task B. For example, a rendezvous object. A sees that somebody (B) is waiting, thus it stores the data, sets the event, and continues on its merry way. B's `receive` call on this object must now return the data that A stored in it.

If that doesn't happen when B gets cancelled, suddenly nobody is responsible for the data in question. You can't fix this by teaching the `send` side to wait for the data to have been read because, surprise, again you'd have to wait on an event, and now you run into the exact same problem on A's side: if it gets cancelled, you don't know whether the owner of the object is now A or B.

While it's certainly possible to add higher-level workarounds like an atomic and non-waiting "get the data out of this object if there happens to be any in there" method, that doesn't match every other `read` or `get` or `receive` call in Trio/AnyIO. The basic guarantee should be to either return the data that's been read, or to leave the channel in the state it was in before the call was made. (Same with `send`/`put`/`write`.)

**Comment by agronholm** [comment]
Don't you think those guarantees go right out the window when tasks start getting cancelled (rather than the memory object streams being closed)? And OP's argument was that, if a task was cancelled after being scheduled, AnyIO should somehow override asyncio's normal cancellation mechanism to allow the task to continue instead of raising a cancellation error like it normally would?

**Comment by smurfix** [comment]
Well they don't go out the window when you talk to network connections, so why should they do so when using a memory stream? (Besides, well that's just the first example I could think of, I'm sure there are others where this may be an issue.)

If fixing / working around this by way of the AnyIO wrapper is too difficult / not possible / degrades performance too much / too much work for too little gain / …, given asyncio's current cancellation semantics, the alternate solution would be to (a) document that yes this is a problem with no generic workaround by (mis)design, use Trio if you need a way to get guaranteed behavior, and (b) teach asyncio sane cancellation semantics so that this gets fixed long term.

**Comment by agronholm** [comment]
> Well they don't go out the window when you talk to network connections, so why should they do so when using a memory stream? (Besides, well that's just the first example I could think of, I'm sure there are others where this may be an issue.)

If you're reading from a socket in a task, and the task is scheduled to be resumed and then gets cancelled, another task could take over and start receiving data from it without anything being lost.

As for memory streams, we already handle this problem ([here](https://github.com/agronholm/anyio/blob/master/src/anyio/streams/memory.py#L107-L111)) by ignoring the first cancellation so as not to lose the data.
